env/env_encrypt2.cc

```diff
@@ -0,0 +1,572 @@
+    }
+
+    ret_val = EncryptedEnv2::crypto_.EVP_CIPHER_CTX_reset(aes_context.get());
+    if (1 == ret_val) {
+      // https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38a.pdf
+      memcpy(iv.bytes, nonce_, AES_BLOCK_SIZE / 2);
+      EncodeFixed64((char*)&iv.bytes[AES_BLOCK_SIZE / 2],
+                    blockIndex);  // this will be little endian
+      block_in.nums[0] = 0;
+      block_in.nums[1] = 0;
+
+      ret_val = EncryptedEnv2::crypto_.EVP_EncryptInit_ex(
+          aes_context.get(), EncryptedEnv2::crypto_.EVP_aes_256_ctr(), nullptr,
+          key_.key, iv.bytes);
+      if (1 == ret_val) {
+        ret_val = EncryptedEnv2::crypto_.EVP_EncryptUpdate(
+            aes_context.get(), block_out.bytes, &out_len, block_in.bytes,
+            in_len);
+
+        if (1 != ret_val || AES_BLOCK_SIZE != out_len) {
+          status = Status::InvalidArgument("EVP_EncryptUpdate failed: ",
+                                           AES_BLOCK_SIZE == out_len
+                                               ? "bad return value"
+                                               : "output length short");
+        }
+      } else {
+        status = Status::InvalidArgument("EVP_EncryptInit_ex failed.");
+      }
+    } else {
+      status = Status::InvalidArgument("EVP_CIPHER_CTX_reset failed.");
+    }
+
+    // XOR data with ciphertext.
+    uint64_t* data_ptr;
+    data_ptr = (uint64_t*)data;
+    *data_ptr ^= block_out.nums[0];
+    data_ptr = (uint64_t*)(data + 8);
+    *data_ptr ^= block_out.nums[1];
+  } else {
+    status = Status::NotSupported(
+        "libcrypto not available for encryption/decryption.");
+  }
+
+  return status;
+}
+
+Status AESBlockAccessCipherStream::DecryptBlock(uint64_t blockIndex, char* data,
+                                                char* scratch) {
+  return EncryptBlock(blockIndex, data, scratch);
+}
+
+Status CTREncryptionProvider2::CreateNewPrefix(const std::string& /*fname*/,
+                                               char* prefix,
+                                               size_t prefixLength) {
+  Status s;
+  if (EncryptedEnv2::crypto_.IsValid()) {
+    if (sizeof(Prefix0) <= prefixLength) {
+      int ret_val;
+
+      Prefix0* pf = {(Prefix0*)prefix};
+      memcpy(pf->key_description_, key_desc_.desc, sizeof(key_desc_.desc));
+      ret_val = EncryptedEnv2::crypto_.RAND_bytes(
+          (unsigned char*)&pf->nonce_,
+          AES_BLOCK_SIZE / 2);  // RAND_poll() to initialize
+      if (1 != ret_val) {
+        s = Status::NotSupported("RAND_bytes failed");
+      }
+    } else {
+      s = Status::NotSupported("Prefix size needs to be 28 or more");
+    }
+  } else {
+    s = Status::NotSupported("RAND_bytes() from libcrypto not available.");
+  }
+
+  return s;
+}
+
+// Returns an Env that encrypts data when stored on disk and decrypts data when
+// read from disk.
+Env* NewEncryptedEnv2(Env* base_env, EncryptedEnv2::ReadKeys encrypt_read,
```

Comment:
Shall we call it `EncryptedEnvV2` instead of `EncryptedEnv2`?

Reply:
Done.